network/gossip/libp2p/topology/topicBasedTopology_test.go

```diff
@@ -223,6 +221,168 @@ func (suite *TopicAwareTopologyTestSuite) TestConnectedness_ClusterChannelID() {
+	require.NoError(suite.T(), err)
+
+	sample, err := top.sampleConnectedGraph(suite.all, nil)
+	require.NoError(suite.T(), err)
+
+	// the LinearFanoutGraphSampler utilizes the LinearFanoutFunc. Hence any sample it makes should have
+	// the size equal to applying LinearFanoutFunc over the original set.
+	expectedFanout := LinearFanoutFunc(len(suite.all))
+	require.Equal(suite.T(), len(sample), expectedFanout)
+
+	// checks sample does not include any duplicate
+	suite.uniquenessCheck(sample)
+}
+
+// TestLinearFanout_ConditionalSampling evaluates that sampling a connected graph fanout with a shouldHave set
+// follows the LinearFanoutFunc, and it also does not contain duplicate element.
+func (suite *TopicAwareTopologyTestSuite) TestLinearFanout_ConditionalSampling() {
+	// samples 10 all into `shouldHave` set.
+	shouldHave := suite.all.Sample(10)
+
+	// creates a topology for the node
+	top, err := NewTopicBasedTopology(suite.all[0].NodeID, suite.state, suite.subMngr[0])
+	require.NoError(suite.T(), err)
+
+	// samples a connected graph of `all` that includes `shouldHave` set.
+	sample, err := top.sampleConnectedGraph(suite.all, shouldHave)
+	require.NoError(suite.T(), err)
+
+	// the LinearFanoutGraphSampler utilizes the LinearFanoutFunc. Hence any sample it makes should have
+	// the size equal to applying LinearFanoutFunc over the original set.
+	expectedFanout := LinearFanoutFunc(len(suite.all))
+	require.Equal(suite.T(), len(sample), expectedFanout)
+
+	// checks sample does not include any duplicate
+	suite.uniquenessCheck(sample)
+
+	// checks inclusion of all shouldHave ones into sample
+	for _, id := range shouldHave {
+		require.Contains(suite.T(), sample, id)
+	}
+}
+
+// TestLinearFanout_SmallerAll evaluates that sampling a connected graph fanout with a shouldHave set
+// that is greater than required fanout, returns the `shouldHave` set instead.
+func (suite *TopicAwareTopologyTestSuite) TestLinearFanout_SmallerAll() {
+	// samples 10 all into 'shouldHave'.
+	shouldHave := suite.all.Sample(10)
+	// samples a smaller component of all with 5 nodes and combines with `shouldHave`
+	smallerAll := suite.all.Filter(filter.Not(filter.In(shouldHave))).Sample(5).Union(shouldHave)
+
+	// creates a topology for the node
+	top, err := NewTopicBasedTopology(suite.all[0].NodeID, suite.state, suite.subMngr[0])
+	require.NoError(suite.T(), err)
+
+	// total size of smallerAll is 15, and it requires a linear fanout of 8 which is less than
+	// size of `shouldHave` set, so the shouldHave itself should return
+	sample, err := top.sampleConnectedGraph(smallerAll, shouldHave)
+	require.NoError(suite.T(), err)
+	require.Equal(suite.T(), len(sample), len(shouldHave))
+	require.ElementsMatch(suite.T(), sample, shouldHave)
+}
+
+// TestLinearFanout_SubsetViolence evaluates that trying to sample a connected graph when `shouldHave`
+// is not a subset of `all` returns an error.
+func (suite *TopicAwareTopologyTestSuite) TestLinearFanout_SubsetViolence() {
```

Comment:
```suggestion
func (suite *TopicAwareTopologyTestSuite) TestLinearFanout_SubsetViolation() {
```

Hopefully there's no violence!